### DISCUSSION TITLE
cli ftp with path completion and automatic anonymous login??

### DISCUSSION THREAD

**Discussion by goyalyashpal** [msys2/MSYS2-packages#3534] [Q&A] [answered]
### Summary:

**Problem:** Want to transfer files from android to pc over LAN (hotspot-wifi)
**Solution:** Looking for a CLI FTP software (FOSS) which essentially is as same as local `bash` experience. 

This boils down to following features:
1. `ls` and `cd` i.e. can list the contents and change directory so that the ftp server can be browsed/explored
2. file path completion suggestion with tab
3. supports anonymous login

So, the imagined user flow will be smth like:

```
$ ftp
ftp> open <address> <port>
ftp> ls
ftp> cd ABC/ <tab - does ls for matching name> / <tab>
ftp> get DE<tab - does ls for matching name>
```

### Details

hi!

* i use ftp to transfer files between my android device and my windows device over local area network (hotspot-wifi)
* i currently use GUI file explorers or GUI internet browsers for that...
* and ftp (in file explorers) behaves almost exactly like local directory, that is, seemless...
* i am looking for similar seemless experience on CLI

But recently had to use a PC where windows explorer was hanging tooooo much, so, can't use that. So, I thought that maybe CLI from MSYS2 has some solution to this :smiley:

util | invocation and starting | entering path and browsing
-- | -- | --
<tt>[wget]</tt> | supports (3) anonymous logins | <ul><li>No (2) or (1)</li><li> ∴ requires manually entering FULL path</li></ul>
GNU <tt>[ftp]</tt> | <ul><li>no (3) anonymous logins, (see [man][man ftp]) ∴ cumbersome</li><li>shadowed by MS `ftp` ∴ long invocation `[/usr]/bin/ftp` </li></ul> | <ul><li>(1) `ls` and `cd` provides at least some <br/> aid in entering the path step-by-step</li><li>lack of (2) hurts</li></ul>

[wget]: https://packages.msys2.org/search?q=wget
[ftp]: https://packages.msys2.org/base/inetutils
[man ftp]: https://www.mankier.com/1/ftp

These were the only 2 I tried so far. Further prospects might be (have not tried):
* SFTP
* SCP or WinSCP (`s`ecured `c`o`p`y)
* Filezilla

### Questions:

1. Does `ftp` or some other CLI software supports path completion suggestions and anonymous login? 
i.e. maybe it requires some config or possible via some function etc.
2. How do you tackle ftp? which client do you use and recommend?

Additionally I will also be interested to hear any other suggestion to transfer files wirelessly from phone to pc? (No bluetooth please 😆)
Like maybe some other protocol (HTTP or TCP) which provides aforementioned CLI features or some other way altogether maybe?

**Comment by Biswa96** [comment]
> Want to transfer files from android to pc over LAN 

You can use ADB (Android Debug Bridge) or KDE Connect.

**Reply by goyalyashpal** [reply]
> ADB

ADB - wireless ADB right? Well, as of now, it is super picky where 
- u can't use the phone's hotspot 
- and ∵ my laptop can't create hotspot of its own unless its connected to a data already (yes, super weird, and escapes my understanding)
- ∴ it requires a 3rd device to create the netword, which i don't have....

> KDE Connect

thanks, will look into it :+1:

**Reply by Biswa96** [reply]
> and ∵ my laptop can't create hotspot of its own 

Didn't you asked about using hotspot-wifi?

**Reply by goyalyashpal** [reply]
hotspot from phone, wifi of laptop

**Comment by goyalyashpal** [comment]
some resources:
* [filezilla CLI] (via "Client Installation" instructions)
* winscp: [cli options][winscp cli options] | [command terminal UI][winscp command ui interface]
* GNU network utils [inetutils manual]: Contents # 11 [FTP][ftp in contents]

[filezilla CLI]: https://wiki.filezilla-project.org/Command-line_arguments_(Client)
[winscp CLI options]: https://winscp.net/eng/docs/commandline
[winscp command ui interface]: https://winscp.net/eng/docs/ui_commander#command_line
[inetutils manual]: https://www.gnu.org/software/inetutils/manual/index.html
[ftp in contents]: https://www.gnu.org/software/inetutils/manual/html_node/index.html#toc-ftp_003a-FTP-client

**Reply by goyalyashpal** [reply]
Found 2 clients with completion features: (ref: article by [TechMint])

1. [yafc] - Yet Another FTP Client, 
    - it is said that it has "tab completion" 
    - it is FOSS.

2. [NcFTP] - see [manual][man ncftp]
    - "filename completion"
    - Freeware proprietary

[techmint]: https://www.tecmint.com/command-line-ftp-clients-for-linux/
[yafc]: https://yafc.sourceforge.net/
[ncftp]: https://www.ncftp.com/
[man ncftp]: https://web.archive.org/web/20210803074725/https://www.mankier.com/1/ncftp#:~:text=no%20longer%20does%20command%2Dline
[cvs]: https://en.wikipedia.org/wiki/Concurrent_Versions_System

the problems with the FOSS yafc are that:
- it is unmaintained (2005) _AND_
- the project was using [CVS].

**Reply by goyalyashpal** [reply]
Due to CVS, this repo doesn't seem to be salvageable to git, at least not by me.... But still...

I thoughto [search] about it: 
* "Migrate a CVS module [from sourceforge] to Git (github)" on [dossy.org]
* SF also has an article on it: [CVS documentation]

[search]: https://duckduckgo.com/?q=importing+a+cvs+repo+from+sourceforge+to+github&t=vivaldi&ia=web&iax=qa
[dossy.org]: https://dossy.org/2015/12/how-to-migrate-a-cvs-module-to-git-and-github/
[CVS documentation]: https://sourceforge.net/p/forge/documentation/CVS/

The tools used (all found on msys2):
- `rsync` (net-utils) 
- `cvs` 
- `git` 
- svn or `subversion` (vcs)

For conversions (none found on msys2):
- <tt>git [cvsimport][gitcvs]</tt> [Requires `cvsps`]
- <tt>[cvs2git]</tt> [Based on <tt>[cvs2svn]</tt> which requires Python 2, doesn't work with Py3]

[cvs2svn]: http://www.mcs.anl.gov/~jacob/cvs2svn/cvs2svn.html
[cvs2git]: http://www.mcs.anl.gov/~jacob/cvs2svn/cvs2git.html#usage
[gitcvs]: https://git-scm.com/docs/gitcvs-migration#_importing_a_cvs_archive

Remarks:

> other importers may provide better results. See the note in [git-cvsimport[1]](https://git-scm.com/docs/git-cvsimport) for other options
\- [gitcvs-migration][gitcvs]
>
> If you are performing a one-shot import of a CVS repository consider using [cvs2git](http://cvs2svn.tigris.org/cvs2git.html) or [cvs-fast-export](http://www.catb.org/esr/cvs-fast-export/).
\- [git-cvsimport[1]](https://git-scm.com/docs/git-cvsimport)



I thought maybe convert to svn, then let the github take it from there (i.e. the SF article) and these converters are not there in msys2 packages... will have to searcn elsewhere on the net for that.... And that Py2 requirement will have to be catered to too.

**Reply by goyalyashpal** [reply]
SO: [migrate-from-cvs-to-git] shares lots of input on this matter, with hugely varying mileage with each tool
But the thing is, I don't have much concern with history of that repo... as I don't have _any_ data whatsoever on it.

aah, it _seems_ hard, i wish the GNU FTP merged it in while they could 😢 


[migrate-from-cvs-to-git]: https://stackoverflow.com/questions/20869710/#28951593

- cvs-fast-export and cvsps are both by same dev Eric Raymond, with _latter_ being deprecated [^cvsps-d]
- _**cvs-fast-export**_ is written in C, so, hard to debug
- for cvs2git or cvs2svn, getting Python2 to run is huge work, or use docker [^dock]
- use reposurgeon by Eric Raymond [^repos]

[^cvsps-d]: cvsps deprecated, cvs-fast-export in C, Python2 environment hard 
https://stackoverflow.com/a/20869792/8395964

[^repos]: Reposurgeon by Eric: 
https://stackoverflow.com/a/43786545/8395964

[^dock]: Use Docker for Py2: 
https://stackoverflow.com/a/72220971/8395964

**Reply by goyalyashpal** [reply]
i remember seeing ncftp mentioned as being used in some FOSS software yesterday, but couldn't recall where....

**Reply by goyalyashpal** [reply]
https://github.com/msys2/MSYS2-packages/issues/3565

**Comment by goyalyashpal** [comment, marked as answer]
I found <tt>[lftp]</tt> - it supports autonomous login, and does have fantastic (but case sensitive) auto completion. 
Niiice. Now I can peacefully use ftp for file transfers between android to pc.


<details>
<summary>How I found this</summary>

i was doing `pacman -Ss man` (lots of good finds there: `z3`, `axel`, etc) and stumbled upon this: <tt>[lftp]</tt>

[lftp]: https://packages.msys2.org/package/lftp

```console
$ pacman -Ss man
msys/lftp 4.9.2-4 (net-utils)
    Sophisticated command line based FTP client (net-utils)
   ...

$ # "Sophisticated" huh! maybe it does support autonomous, and tab completion.
$ pacman -Si lftp
URL             : https://lftp.yar.ru/
   ...

$ open https://lftp.yar.ru/
```

I searched there for "tab" and found this: [remote-completion]. Nice. 
Therefore, gave it an install, and tried. it worked awesome. :smiley:

</details>

<details>
<summary>Usage notes to self:</summary><br/>

On pc side:
- Its autocompletion with `'` escaping for space containing names seems to work even better than `bash`
- Experience from `ftp` usage helped quite a lot. `-p` option can be given in end too: `lftp <site> -p <port>` 

On android side:
- MiXplorer v6.58.8's "Read-only" option in "Accounts" in Server settings doesn't work. 
- "[Material Files]" v1.5.2 "Allow writing" toggle works.

</details>

[remote-completion]: https://lftp.yar.ru/lftp-man.html#:~:text=completion.%20When%20%20true%2C-,Tab,-key%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20guesses%20if
[Material Files]: https://f-droid.org/packages/me.zhanghai.android.files

</details>
